### PR TITLE
perf: on-demand sourcemap generation

### PR DESCRIPTION
**Pull request by h-a-n-a**
## Checklist

- [x] Tests updated (or not required).
- [x] Documentation updated (or not required).

### CODE REVIEW

**Comment by h-a-n-a** [comment]
!bench

**Comment by h-a-n-a** [comment]
!bench

**Comment by h-a-n-a** [comment]
!bench

File: crates/rspack_loader_swc/src/lib.rs [hunk 1]
```diff
@@ -122,9 +122,9 @@ impl Loader<LoaderRunnerContext> for SwcLoader {
         .as_ref()
         .map(|v| matches!(v, OutputCharset::Ascii)),
       source_map_config: SourceMapConfig {
-        enable: !matches!(source_map_kind, SourceMapKind::None),
-        inline_sources_content: true,
-        emit_columns: matches!(source_map_kind, SourceMapKind::SourceMap),
+        enable: source_map_kind.source_map(),
```
Comment (h-a-n-a) [inline crates/rspack_loader_swc/src/lib.rs:125]: Only generate sourcemap if it's `module`.

File: crates/rspack_loader_swc/src/lib.rs [hunk 2]
```diff
@@ -122,9 +122,9 @@ impl Loader<LoaderRunnerContext> for SwcLoader {
...
+        inline_sources_content: source_map_kind.source_map(),
+        emit_columns: !source_map_kind.cheap(),
```
Comment (h-a-n-a) [inline crates/rspack_loader_swc/src/lib.rs:127]: Do not emit columns if it's `cheap`.

**Comment by h-a-n-a** [comment]
!bench

**Review by SyMind — APPROVED** [review]

**Comment by pushkargupta9807** [comment]
can we please have more info on this perf improvement PR  ? For instance how can we use it , examples etc

**Comment by chenjiahan** [comment]
@h-a-n-a please check this ⬆️

**Comment by h-a-n-a** [comment]
> can we please have more info on this perf improvement PR ? For instance how can we use it , examples etc

This is an internal optimisation where source maps should always be generated on-demand by `builtin:swc-loader` with the respect to the compiler's source map options. I didn't make a real world benchmark, but this will improve the case where `options.devtool` set to `false`.

**Pull request merged**